{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "7e442b63-bbdc-41e7-a34e-960038391c05",
   "metadata": {},
   "source": [
    "Computing stats for single cell RNA bubbleplot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d9b78890-c7d6-4046-aab4-cb993c9f55b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import anndata as ad\n",
    "import synapseclient"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c4a403f5-b465-4cef-a3f3-b17f67578636",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "UPGRADE AVAILABLE\n",
      "\n",
      "A more recent version of the Synapse Client (4.2.0) is available. Your version (2.7.2) can be upgraded by typing:\n",
      "    pip install --upgrade synapseclient\n",
      "\n",
      "Python Synapse Client version 4.2.0 release notes\n",
      "\n",
      "https://python-docs.synapse.org/news/\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Welcome, heimann!\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 342749 × 36601\n",
       "    obs: 'n_genes_by_counts', 'log1p_n_genes_by_counts', 'total_counts', 'log1p_total_counts', 'pct_counts_in_top_50_genes', 'pct_counts_in_top_100_genes', 'pct_counts_in_top_200_genes', 'pct_counts_in_top_500_genes', 'percent_mito', 'n_counts', 'percent_ribo', 'scrublet', 'batch', 'CD45_enrich', 'batch_num', 'cohort', 'pCR', 'RCB', 'cleared_nodes', 'treatment', 'patient_treatment', 'hormone_receptor', 'combined_tcr', 'umap1', 'umap2', 'leiden_50nbr_res1.2', 'celltype', 'global_clusters', 'bcell_leiden_nbr100_res0.6', 'tcell_leiden_nbr100_res0.6', 'myeloid_leiden_nbr30_res0.8', 'subcluster'\n",
       "    var: 'gene_ids', 'feature_types'\n",
       "    uns: 'bcell_leiden_nbr100_res0.6_colors', 'celltype_colors', 'global_clusters_colors', 'leiden_50nbr_res1.2_colors', 'log1p', 'myeloid_leiden_nbr30_res0.8_colors', 'subcluster_colors', 'tcell_leiden_nbr100_res0.6_colors'\n",
       "    obsm: 'X_bcell_umap', 'X_myeloid_umap', 'X_tcell_umap', 'X_umap'\n",
       "    layers: 'counts', 'normalized'"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#loading h5ad file after qc, described in 'msk_data_prep.ipynb'\n",
    "syn = synapseclient.Synapse()\n",
    "syn.login()\n",
    "\n",
    "#Load data\n",
    "entity = syn.get('syn55258687')\n",
    "adata = sc.read_h5ad(entity.path)\n",
    "\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "94a4b39c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_ids</th>\n",
       "      <th>feature_types</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MIR1302-2HG</th>\n",
       "      <td>ENSG00000243485</td>\n",
       "      <td>Gene Expression</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FAM138A</th>\n",
       "      <td>ENSG00000237613</td>\n",
       "      <td>Gene Expression</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OR4F5</th>\n",
       "      <td>ENSG00000186092</td>\n",
       "      <td>Gene Expression</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AL627309.1</th>\n",
       "      <td>ENSG00000238009</td>\n",
       "      <td>Gene Expression</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AL627309.3</th>\n",
       "      <td>ENSG00000239945</td>\n",
       "      <td>Gene Expression</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC141272.1</th>\n",
       "      <td>ENSG00000277836</td>\n",
       "      <td>Gene Expression</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC023491.2</th>\n",
       "      <td>ENSG00000278633</td>\n",
       "      <td>Gene Expression</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC007325.1</th>\n",
       "      <td>ENSG00000276017</td>\n",
       "      <td>Gene Expression</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC007325.4</th>\n",
       "      <td>ENSG00000278817</td>\n",
       "      <td>Gene Expression</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC007325.2</th>\n",
       "      <td>ENSG00000277196</td>\n",
       "      <td>Gene Expression</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>36601 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    gene_ids    feature_types\n",
       "MIR1302-2HG  ENSG00000243485  Gene Expression\n",
       "FAM138A      ENSG00000237613  Gene Expression\n",
       "OR4F5        ENSG00000186092  Gene Expression\n",
       "AL627309.1   ENSG00000238009  Gene Expression\n",
       "AL627309.3   ENSG00000239945  Gene Expression\n",
       "...                      ...              ...\n",
       "AC141272.1   ENSG00000277836  Gene Expression\n",
       "AC023491.2   ENSG00000278633  Gene Expression\n",
       "AC007325.1   ENSG00000276017  Gene Expression\n",
       "AC007325.4   ENSG00000278817  Gene Expression\n",
       "AC007325.2   ENSG00000277196  Gene Expression\n",
       "\n",
       "[36601 rows x 2 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "7542f94a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#saving the expression matrix\n",
    "from scipy.sparse import csr_matrix, find\n",
    "A = adata.layers['normalized']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "0d188d12",
   "metadata": {},
   "outputs": [],
   "source": [
    "#get non-zero elements\n",
    "rows, cols = A.nonzero()\n",
    "values = A.data\n",
    "\n",
    "#organize in dataframe\n",
    "df = pd.DataFrame({\n",
    "    'Row': [(adata.obs).index[row] for row in rows],\n",
    "    'Column': [adata.var.index[col] for col in cols],\n",
    "    'Value': values\n",
    "})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "99d1ec0d",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_csv('Shiao_2024_norm_expr.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "5c82f640",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "##################################################\n",
      " Uploading file to Synapse storage \n",
      "##################################################\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "File(modifiedOn='2024-05-06T22:35:09.504Z', createdOn='2024-05-06T22:35:09.358Z', path='Shiao_2024_norm_expr.tsv', files=['Shiao_2024_norm_expr.tsv'], modifiedBy='3398555', versionLabel='1', synapseStore=True, isLatestVersion=True, dataFileHandleId='140385388', parentId='syn55271493', etag='689a90cf-5cca-4c31-9e51-91d9f4574373', createdBy='3398555', concreteType='org.sagebionetworks.repo.model.FileEntity', cacheDir='', name='Shiao_2024_norm_expr.tsv', versionNumber=1, _file_handle={'id': '140385388', 'etag': 'b51cc4ed-3995-4f16-82d6-0d22e6f1fa4b', 'createdBy': '3398555', 'createdOn': '2024-05-06T22:35:09.000Z', 'modifiedOn': '2024-05-06T22:35:09.000Z', 'concreteType': 'org.sagebionetworks.repo.model.file.S3FileHandle', 'contentType': 'text/tab-separated-values', 'contentMd5': '50609196f5b35213d46c652f8012b61f', 'fileName': 'Shiao_2024_norm_expr.tsv', 'storageLocationId': 1, 'contentSize': 19921242607, 'status': 'AVAILABLE', 'bucketName': 'proddata.sagebase.org', 'key': '3398555/7642cd04-d52d-4d4c-b367-b80a46c33b94/Shiao_2024_norm_expr.tsv', 'isPreview': False, 'externalURL': None}, id='syn59059670')"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "file_entity = synapseclient.File(\"Shiao_2024_norm_expr.tsv\", parent = \"syn55271493\")\n",
    "syn.store(file_entity)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "84496d40-2783-4b3b-b8b2-01562a1b7644",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'syn' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_3453275/3909461930.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;31m#getting the list of immunomodulators from CRI iAtlas\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0mentity\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msyn\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'syn59016496'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m \u001b[0mimmunomodulators\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread_csv\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mentity\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpath\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0mimmunomodulators\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'syn' is not defined"
     ]
    }
   ],
   "source": [
    "#getting the list of immunomodulators from CRI iAtlas\n",
    "entity = syn.get('syn59016496')\n",
    "immunomodulators = pd.read_csv(entity.path)\n",
    "\n",
    "immunomodulators"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "eb50866c-6fcf-4ea3-8865-68abf191a79b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0     TNFRSF18\n",
       "1      TNFRSF4\n",
       "2     TNFRSF14\n",
       "3      TNFRSF9\n",
       "4        VTCN1\n",
       "        ...   \n",
       "70        CD40\n",
       "71      ICOSLG\n",
       "72       ITGB2\n",
       "73     ADORA2A\n",
       "74      CD40LG\n",
       "Length: 75, dtype: object"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#get a list of the genes that are present in the dataset & are immunomodulators\n",
    "genes = pd.Series(adata.var[np.isin(adata.var.index, immunomodulators[\"hgnc\"])].index)\n",
    "genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "a35d88a7-da41-4dd4-8286-91489ee2d913",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Tcell       188098\n",
       "myeloid     110168\n",
       "Bcell        42841\n",
       "mastcell      1642\n",
       "Name: celltype, dtype: int64"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#compute how many counts map to each cell type\n",
    "freq_cell_types = adata.obs[\"celltype\"].value_counts()\n",
    "freq_cell_types"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "33e07975-3980-4fe2-836a-60263b31dff4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Tcell</td>\n",
       "      <td>TNFRSF18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Tcell</td>\n",
       "      <td>TNFRSF4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Tcell</td>\n",
       "      <td>TNFRSF14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Tcell</td>\n",
       "      <td>TNFRSF9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Tcell</td>\n",
       "      <td>VTCN1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>mastcell</td>\n",
       "      <td>CD40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>mastcell</td>\n",
       "      <td>ICOSLG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>mastcell</td>\n",
       "      <td>ITGB2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>mastcell</td>\n",
       "      <td>ADORA2A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>mastcell</td>\n",
       "      <td>CD40LG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>300 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        cell      gene\n",
       "0      Tcell  TNFRSF18\n",
       "1      Tcell   TNFRSF4\n",
       "2      Tcell  TNFRSF14\n",
       "3      Tcell   TNFRSF9\n",
       "4      Tcell     VTCN1\n",
       "..       ...       ...\n",
       "70  mastcell      CD40\n",
       "71  mastcell    ICOSLG\n",
       "72  mastcell     ITGB2\n",
       "73  mastcell   ADORA2A\n",
       "74  mastcell    CD40LG\n",
       "\n",
       "[300 rows x 2 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#create grid with all cell x gene combinations\n",
    "lp1, lp2 = pd.core.reshape.util.cartesian_product([freq_cell_types.index, genes])\n",
    "cell_gene = pd.DataFrame(dict(cell=lp1, gene=lp2))\n",
    "cell_gene"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e50d4ca0-752f-4eeb-b821-e3049b13963c",
   "metadata": {},
   "source": [
    "We need to compute:\n",
    "- % of cell of a given type that expresses a gene (count expr by type/freq of cell type)\n",
    "- mean expr value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "b92adb7b-ec5d-4703-a98d-872da782ca0b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_expr_by_cell(cell_type, gene):\n",
    "    expr = adata[adata.obs.celltype == cell_type].to_df(layer=\"normalized\")[gene]\n",
    "    counts = sum(expr != 0)\n",
    "    if(counts>0): avg = (expr[expr != 0]).mean()\n",
    "    else: avg = 0 #technically this is wrong, but plotting libraries crash with NAs or characters\n",
    "    return counts, avg"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "bc8f0ea7-2d79-45db-9805-8b9a76f141a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to apply to each row\n",
    "def apply_function(row):\n",
    "    counts, avg = get_expr_by_cell(row['cell'], row['gene'])  \n",
    "    return pd.Series({'counts': counts, 'avg': avg})\n",
    "\n",
    "# Apply the function to each row and concatenate the results\n",
    "#result = pd.concat([cell_gene, cell_gene.apply(apply_function, axis=1)], axis=1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "bd1ec741-f57c-4785-bc46-e5617884d505",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell</th>\n",
       "      <th>gene</th>\n",
       "      <th>counts</th>\n",
       "      <th>avg</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Tcell</td>\n",
       "      <td>TNFRSF18</td>\n",
       "      <td>51316.0</td>\n",
       "      <td>1.998775</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Tcell</td>\n",
       "      <td>TNFRSF4</td>\n",
       "      <td>40241.0</td>\n",
       "      <td>1.986621</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Tcell</td>\n",
       "      <td>TNFRSF14</td>\n",
       "      <td>57782.0</td>\n",
       "      <td>1.612135</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Tcell</td>\n",
       "      <td>TNFRSF9</td>\n",
       "      <td>20166.0</td>\n",
       "      <td>1.555380</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Tcell</td>\n",
       "      <td>VTCN1</td>\n",
       "      <td>293.0</td>\n",
       "      <td>1.691238</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Tcell</td>\n",
       "      <td>SLAMF7</td>\n",
       "      <td>11676.0</td>\n",
       "      <td>1.415022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Tcell</td>\n",
       "      <td>SELP</td>\n",
       "      <td>798.0</td>\n",
       "      <td>1.298185</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Tcell</td>\n",
       "      <td>TNFSF4</td>\n",
       "      <td>6237.0</td>\n",
       "      <td>1.475875</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Tcell</td>\n",
       "      <td>IL10</td>\n",
       "      <td>3136.0</td>\n",
       "      <td>1.695357</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Tcell</td>\n",
       "      <td>IL1A</td>\n",
       "      <td>242.0</td>\n",
       "      <td>1.373145</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    cell      gene   counts       avg\n",
       "0  Tcell  TNFRSF18  51316.0  1.998775\n",
       "1  Tcell   TNFRSF4  40241.0  1.986621\n",
       "2  Tcell  TNFRSF14  57782.0  1.612135\n",
       "3  Tcell   TNFRSF9  20166.0  1.555380\n",
       "4  Tcell     VTCN1    293.0  1.691238\n",
       "5  Tcell    SLAMF7  11676.0  1.415022\n",
       "6  Tcell      SELP    798.0  1.298185\n",
       "7  Tcell    TNFSF4   6237.0  1.475875\n",
       "8  Tcell      IL10   3136.0  1.695357\n",
       "9  Tcell      IL1A    242.0  1.373145"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result = pd.concat([cell_gene, cell_gene.apply(apply_function, axis=1)], axis=1)\n",
    "result.iloc[0:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "eb373162-f00e-4a64-9215-cb5934aa4c59",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell</th>\n",
       "      <th>gene</th>\n",
       "      <th>counts</th>\n",
       "      <th>avg</th>\n",
       "      <th>celltype</th>\n",
       "      <th>perc_expr</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Tcell</td>\n",
       "      <td>TNFRSF18</td>\n",
       "      <td>51316.0</td>\n",
       "      <td>1.998775</td>\n",
       "      <td>188098</td>\n",
       "      <td>0.272815</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Tcell</td>\n",
       "      <td>TNFRSF4</td>\n",
       "      <td>40241.0</td>\n",
       "      <td>1.986621</td>\n",
       "      <td>188098</td>\n",
       "      <td>0.213936</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Tcell</td>\n",
       "      <td>TNFRSF14</td>\n",
       "      <td>57782.0</td>\n",
       "      <td>1.612135</td>\n",
       "      <td>188098</td>\n",
       "      <td>0.307191</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Tcell</td>\n",
       "      <td>TNFRSF9</td>\n",
       "      <td>20166.0</td>\n",
       "      <td>1.555380</td>\n",
       "      <td>188098</td>\n",
       "      <td>0.107210</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Tcell</td>\n",
       "      <td>VTCN1</td>\n",
       "      <td>293.0</td>\n",
       "      <td>1.691238</td>\n",
       "      <td>188098</td>\n",
       "      <td>0.001558</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Tcell</td>\n",
       "      <td>SLAMF7</td>\n",
       "      <td>11676.0</td>\n",
       "      <td>1.415022</td>\n",
       "      <td>188098</td>\n",
       "      <td>0.062074</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Tcell</td>\n",
       "      <td>SELP</td>\n",
       "      <td>798.0</td>\n",
       "      <td>1.298185</td>\n",
       "      <td>188098</td>\n",
       "      <td>0.004242</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Tcell</td>\n",
       "      <td>TNFSF4</td>\n",
       "      <td>6237.0</td>\n",
       "      <td>1.475875</td>\n",
       "      <td>188098</td>\n",
       "      <td>0.033158</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Tcell</td>\n",
       "      <td>IL10</td>\n",
       "      <td>3136.0</td>\n",
       "      <td>1.695357</td>\n",
       "      <td>188098</td>\n",
       "      <td>0.016672</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Tcell</td>\n",
       "      <td>IL1A</td>\n",
       "      <td>242.0</td>\n",
       "      <td>1.373145</td>\n",
       "      <td>188098</td>\n",
       "      <td>0.001287</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    cell      gene   counts       avg  celltype  perc_expr\n",
       "0  Tcell  TNFRSF18  51316.0  1.998775    188098   0.272815\n",
       "1  Tcell   TNFRSF4  40241.0  1.986621    188098   0.213936\n",
       "2  Tcell  TNFRSF14  57782.0  1.612135    188098   0.307191\n",
       "3  Tcell   TNFRSF9  20166.0  1.555380    188098   0.107210\n",
       "4  Tcell     VTCN1    293.0  1.691238    188098   0.001558\n",
       "5  Tcell    SLAMF7  11676.0  1.415022    188098   0.062074\n",
       "6  Tcell      SELP    798.0  1.298185    188098   0.004242\n",
       "7  Tcell    TNFSF4   6237.0  1.475875    188098   0.033158\n",
       "8  Tcell      IL10   3136.0  1.695357    188098   0.016672\n",
       "9  Tcell      IL1A    242.0  1.373145    188098   0.001287"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Now we compute the % of cells from a given cell type that have expression for a gene\n",
    "#result = pd.merge(result, freq_cell_types, left_on='cell', right_index=True)\n",
    "result['perc_expr'] = result['counts'] / result['celltype']\n",
    "result.iloc[0:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "7e6d87fd-de6e-45d0-92f6-d79c654c43da",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell</th>\n",
       "      <th>gene</th>\n",
       "      <th>counts</th>\n",
       "      <th>avg</th>\n",
       "      <th>Freq</th>\n",
       "      <th>perc_expr</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Tcell</td>\n",
       "      <td>TNFRSF18</td>\n",
       "      <td>51316.0</td>\n",
       "      <td>1.998775</td>\n",
       "      <td>188098</td>\n",
       "      <td>0.272815</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Tcell</td>\n",
       "      <td>TNFRSF4</td>\n",
       "      <td>40241.0</td>\n",
       "      <td>1.986621</td>\n",
       "      <td>188098</td>\n",
       "      <td>0.213936</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Tcell</td>\n",
       "      <td>TNFRSF14</td>\n",
       "      <td>57782.0</td>\n",
       "      <td>1.612135</td>\n",
       "      <td>188098</td>\n",
       "      <td>0.307191</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Tcell</td>\n",
       "      <td>TNFRSF9</td>\n",
       "      <td>20166.0</td>\n",
       "      <td>1.555380</td>\n",
       "      <td>188098</td>\n",
       "      <td>0.107210</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Tcell</td>\n",
       "      <td>VTCN1</td>\n",
       "      <td>293.0</td>\n",
       "      <td>1.691238</td>\n",
       "      <td>188098</td>\n",
       "      <td>0.001558</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Tcell</td>\n",
       "      <td>SLAMF7</td>\n",
       "      <td>11676.0</td>\n",
       "      <td>1.415022</td>\n",
       "      <td>188098</td>\n",
       "      <td>0.062074</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Tcell</td>\n",
       "      <td>SELP</td>\n",
       "      <td>798.0</td>\n",
       "      <td>1.298185</td>\n",
       "      <td>188098</td>\n",
       "      <td>0.004242</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Tcell</td>\n",
       "      <td>TNFSF4</td>\n",
       "      <td>6237.0</td>\n",
       "      <td>1.475875</td>\n",
       "      <td>188098</td>\n",
       "      <td>0.033158</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Tcell</td>\n",
       "      <td>IL10</td>\n",
       "      <td>3136.0</td>\n",
       "      <td>1.695357</td>\n",
       "      <td>188098</td>\n",
       "      <td>0.016672</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Tcell</td>\n",
       "      <td>IL1A</td>\n",
       "      <td>242.0</td>\n",
       "      <td>1.373145</td>\n",
       "      <td>188098</td>\n",
       "      <td>0.001287</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    cell      gene   counts       avg    Freq  perc_expr\n",
       "0  Tcell  TNFRSF18  51316.0  1.998775  188098   0.272815\n",
       "1  Tcell   TNFRSF4  40241.0  1.986621  188098   0.213936\n",
       "2  Tcell  TNFRSF14  57782.0  1.612135  188098   0.307191\n",
       "3  Tcell   TNFRSF9  20166.0  1.555380  188098   0.107210\n",
       "4  Tcell     VTCN1    293.0  1.691238  188098   0.001558\n",
       "5  Tcell    SLAMF7  11676.0  1.415022  188098   0.062074\n",
       "6  Tcell      SELP    798.0  1.298185  188098   0.004242\n",
       "7  Tcell    TNFSF4   6237.0  1.475875  188098   0.033158\n",
       "8  Tcell      IL10   3136.0  1.695357  188098   0.016672\n",
       "9  Tcell      IL1A    242.0  1.373145  188098   0.001287"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result.columns = [\"cell\", \"gene\", \"counts\", \"avg\", \"Freq\", \"perc_expr\"]\n",
    "result.iloc[0:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "915f8f13-07e2-4eef-948f-4701825a0baa",
   "metadata": {},
   "outputs": [],
   "source": [
    "#The Shiao data has different names for cells than what we will use in iAtlas, so let's update them\n",
    "result['cell'] = result['cell'].replace(['Tcell', 'myeloid', 'Bcell', 'mastcell'], ['T cell', 'myeloid cell', 'B cell', 'mast cell'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "518971f3-2514-400c-8301-3f532302d495",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Add dataset info and save data into file\n",
    "result['dataset'] = \"Shiao_2024\"\n",
    "result.to_csv('Shiao_2024_bubble_plot_df.tsv', sep='\\t', index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
