{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "3df40e33",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import synapseclient\n",
    "import matplotlib.pyplot as plt\n",
    "import scipy.sparse as sps\n",
    "from scipy.io import mmread\n",
    "import anndata as ad"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eb109ee8",
   "metadata": {},
   "source": [
    "# Bi 2021 dataset\n",
    "Source: https://www.cell.com/cancer-cell/fulltext/S1535-6108(21)00117-3\n",
    "\n",
    "Stored at: https://singlecell.broadinstitute.org/single_cell/study/SCP1288/tumor-and-immune-reprogramming-during-immunotherapy-in-advanced-renal-cell-carcinoma#study-summary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ff934657-af65-41c9-b0e3-481db3c046fc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Welcome, heimann!\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "UPGRADE AVAILABLE\n",
      "\n",
      "A more recent version of the Synapse Client (4.3.0) is available. Your version (2.7.2) can be upgraded by typing:\n",
      "    pip install --upgrade synapseclient\n",
      "\n",
      "Python Synapse Client version 4.3.0 release notes\n",
      "\n",
      "https://python-docs.synapse.org/news/\n",
      "\n"
     ]
    }
   ],
   "source": [
    "syn = synapseclient.Synapse()\n",
    "syn.login()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "829e78b9-c0ad-43b3-9bd8-183bc3b18480",
   "metadata": {},
   "source": [
    "## Loading data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e0af580e-fe51-46cc-95e4-e84eaacdfb1c",
   "metadata": {},
   "source": [
    "### Clinical Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "e80b8d16",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[WARNING] /home/kawakami/miniconda3/envs/pvision/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3457: DtypeWarning: Columns (14) have mixed types.Specify dtype option on import or set low_memory=False.\n",
      "  exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>NAME</th>\n",
       "      <th>biosample_id</th>\n",
       "      <th>donor_id</th>\n",
       "      <th>species</th>\n",
       "      <th>species__ontology_label</th>\n",
       "      <th>disease</th>\n",
       "      <th>disease__ontology_label</th>\n",
       "      <th>organ</th>\n",
       "      <th>organ__ontology_label</th>\n",
       "      <th>library_preparation_protocol</th>\n",
       "      <th>library_preparation_protocol__ontology_label</th>\n",
       "      <th>ICB_Exposed</th>\n",
       "      <th>ICB_Response</th>\n",
       "      <th>TKI_Exposed</th>\n",
       "      <th>Initial_Louvain_Cluster</th>\n",
       "      <th>Lineage</th>\n",
       "      <th>InferCNV</th>\n",
       "      <th>FinalCellType</th>\n",
       "      <th>sex</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TYPE</td>\n",
       "      <td>group</td>\n",
       "      <td>group</td>\n",
       "      <td>group</td>\n",
       "      <td>group</td>\n",
       "      <td>group</td>\n",
       "      <td>group</td>\n",
       "      <td>group</td>\n",
       "      <td>group</td>\n",
       "      <td>group</td>\n",
       "      <td>group</td>\n",
       "      <td>group</td>\n",
       "      <td>group</td>\n",
       "      <td>group</td>\n",
       "      <td>group</td>\n",
       "      <td>group</td>\n",
       "      <td>group</td>\n",
       "      <td>group</td>\n",
       "      <td>group</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAACCTGAGAATAGGG.p55</td>\n",
       "      <td>P55_scRNA</td>\n",
       "      <td>P55</td>\n",
       "      <td>NCBITaxon_9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>MONDO_0005005</td>\n",
       "      <td>clear cell renal carcinoma</td>\n",
       "      <td>UBERON_0000916</td>\n",
       "      <td>abdomen</td>\n",
       "      <td>EFO:0009899</td>\n",
       "      <td>10X 3' v2 sequencing</td>\n",
       "      <td>ICB</td>\n",
       "      <td>ICB_PR</td>\n",
       "      <td>TKI</td>\n",
       "      <td>9</td>\n",
       "      <td>Lymphoid</td>\n",
       "      <td>NonMalignant</td>\n",
       "      <td>41BB-Hi CD8+ T cell</td>\n",
       "      <td>male</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAACCTGAGGCTAGGT.p55</td>\n",
       "      <td>P55_scRNA</td>\n",
       "      <td>P55</td>\n",
       "      <td>NCBITaxon_9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>MONDO_0005005</td>\n",
       "      <td>clear cell renal carcinoma</td>\n",
       "      <td>UBERON_0000916</td>\n",
       "      <td>abdomen</td>\n",
       "      <td>EFO:0009899</td>\n",
       "      <td>10X 3' v2 sequencing</td>\n",
       "      <td>ICB</td>\n",
       "      <td>ICB_PR</td>\n",
       "      <td>TKI</td>\n",
       "      <td>9</td>\n",
       "      <td>Lymphoid</td>\n",
       "      <td>NonMalignant</td>\n",
       "      <td>41BB-Hi CD8+ T cell</td>\n",
       "      <td>male</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAACCTGCACTGTGTA.p55</td>\n",
       "      <td>P55_scRNA</td>\n",
       "      <td>P55</td>\n",
       "      <td>NCBITaxon_9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>MONDO_0005005</td>\n",
       "      <td>clear cell renal carcinoma</td>\n",
       "      <td>UBERON_0000916</td>\n",
       "      <td>abdomen</td>\n",
       "      <td>EFO:0009899</td>\n",
       "      <td>10X 3' v2 sequencing</td>\n",
       "      <td>ICB</td>\n",
       "      <td>ICB_PR</td>\n",
       "      <td>TKI</td>\n",
       "      <td>9</td>\n",
       "      <td>Lymphoid</td>\n",
       "      <td>NonMalignant</td>\n",
       "      <td>41BB-Hi CD8+ T cell</td>\n",
       "      <td>male</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAACCTGCAGTCCTTC.p55</td>\n",
       "      <td>P55_scRNA</td>\n",
       "      <td>P55</td>\n",
       "      <td>NCBITaxon_9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>MONDO_0005005</td>\n",
       "      <td>clear cell renal carcinoma</td>\n",
       "      <td>UBERON_0000916</td>\n",
       "      <td>abdomen</td>\n",
       "      <td>EFO:0009899</td>\n",
       "      <td>10X 3' v2 sequencing</td>\n",
       "      <td>ICB</td>\n",
       "      <td>ICB_PR</td>\n",
       "      <td>TKI</td>\n",
       "      <td>2</td>\n",
       "      <td>Lymphoid</td>\n",
       "      <td>NonMalignant</td>\n",
       "      <td>MitoHigh T-Helper</td>\n",
       "      <td>male</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34322</th>\n",
       "      <td>TTGGAACGTGAGGGAG.p916</td>\n",
       "      <td>P916_scRNA</td>\n",
       "      <td>P916</td>\n",
       "      <td>NCBITaxon_9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>MONDO_0005005</td>\n",
       "      <td>clear cell renal carcinoma</td>\n",
       "      <td>UBERON_0000029</td>\n",
       "      <td>lymph node</td>\n",
       "      <td>EFO:0009899</td>\n",
       "      <td>10X 3' v2 sequencing</td>\n",
       "      <td>NoICB</td>\n",
       "      <td>NoICB</td>\n",
       "      <td>NoTKI</td>\n",
       "      <td>7</td>\n",
       "      <td>Lymphoid</td>\n",
       "      <td>NonMalignant</td>\n",
       "      <td>Effector T-Helper</td>\n",
       "      <td>male</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34323</th>\n",
       "      <td>TTGTAGGGTATGAAAC.p916</td>\n",
       "      <td>P916_scRNA</td>\n",
       "      <td>P916</td>\n",
       "      <td>NCBITaxon_9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>MONDO_0005005</td>\n",
       "      <td>clear cell renal carcinoma</td>\n",
       "      <td>UBERON_0000029</td>\n",
       "      <td>lymph node</td>\n",
       "      <td>EFO:0009899</td>\n",
       "      <td>10X 3' v2 sequencing</td>\n",
       "      <td>NoICB</td>\n",
       "      <td>NoICB</td>\n",
       "      <td>NoTKI</td>\n",
       "      <td>0</td>\n",
       "      <td>Lymphoid</td>\n",
       "      <td>NonMalignant</td>\n",
       "      <td>41BB-Lo CD8+ T cell</td>\n",
       "      <td>male</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34324</th>\n",
       "      <td>TTTACTGCACACATGT.p916</td>\n",
       "      <td>P916_scRNA</td>\n",
       "      <td>P916</td>\n",
       "      <td>NCBITaxon_9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>MONDO_0005005</td>\n",
       "      <td>clear cell renal carcinoma</td>\n",
       "      <td>UBERON_0000029</td>\n",
       "      <td>lymph node</td>\n",
       "      <td>EFO:0009899</td>\n",
       "      <td>10X 3' v2 sequencing</td>\n",
       "      <td>NoICB</td>\n",
       "      <td>NoICB</td>\n",
       "      <td>NoTKI</td>\n",
       "      <td>9</td>\n",
       "      <td>Lymphoid</td>\n",
       "      <td>NonMalignant</td>\n",
       "      <td>41BB-Hi CD8+ T cell</td>\n",
       "      <td>male</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34325</th>\n",
       "      <td>TTTGTCAAGAGCAATT.p916</td>\n",
       "      <td>P916_scRNA</td>\n",
       "      <td>P916</td>\n",
       "      <td>NCBITaxon_9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>MONDO_0005005</td>\n",
       "      <td>clear cell renal carcinoma</td>\n",
       "      <td>UBERON_0000029</td>\n",
       "      <td>lymph node</td>\n",
       "      <td>EFO:0009899</td>\n",
       "      <td>10X 3' v2 sequencing</td>\n",
       "      <td>NoICB</td>\n",
       "      <td>NoICB</td>\n",
       "      <td>NoTKI</td>\n",
       "      <td>7</td>\n",
       "      <td>Lymphoid</td>\n",
       "      <td>NonMalignant</td>\n",
       "      <td>Effector T-Helper</td>\n",
       "      <td>male</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34326</th>\n",
       "      <td>TTTGTCAAGCGTTTAC.p916</td>\n",
       "      <td>P916_scRNA</td>\n",
       "      <td>P916</td>\n",
       "      <td>NCBITaxon_9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>MONDO_0005005</td>\n",
       "      <td>clear cell renal carcinoma</td>\n",
       "      <td>UBERON_0000029</td>\n",
       "      <td>lymph node</td>\n",
       "      <td>EFO:0009899</td>\n",
       "      <td>10X 3' v2 sequencing</td>\n",
       "      <td>NoICB</td>\n",
       "      <td>NoICB</td>\n",
       "      <td>NoTKI</td>\n",
       "      <td>9</td>\n",
       "      <td>Lymphoid</td>\n",
       "      <td>NonMalignant</td>\n",
       "      <td>41BB-Lo CD8+ T cell</td>\n",
       "      <td>male</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>34327 rows × 19 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        NAME biosample_id donor_id         species  \\\n",
       "0                       TYPE        group    group           group   \n",
       "1       AAACCTGAGAATAGGG.p55    P55_scRNA      P55  NCBITaxon_9606   \n",
       "2       AAACCTGAGGCTAGGT.p55    P55_scRNA      P55  NCBITaxon_9606   \n",
       "3       AAACCTGCACTGTGTA.p55    P55_scRNA      P55  NCBITaxon_9606   \n",
       "4       AAACCTGCAGTCCTTC.p55    P55_scRNA      P55  NCBITaxon_9606   \n",
       "...                      ...          ...      ...             ...   \n",
       "34322  TTGGAACGTGAGGGAG.p916   P916_scRNA     P916  NCBITaxon_9606   \n",
       "34323  TTGTAGGGTATGAAAC.p916   P916_scRNA     P916  NCBITaxon_9606   \n",
       "34324  TTTACTGCACACATGT.p916   P916_scRNA     P916  NCBITaxon_9606   \n",
       "34325  TTTGTCAAGAGCAATT.p916   P916_scRNA     P916  NCBITaxon_9606   \n",
       "34326  TTTGTCAAGCGTTTAC.p916   P916_scRNA     P916  NCBITaxon_9606   \n",
       "\n",
       "      species__ontology_label        disease     disease__ontology_label  \\\n",
       "0                       group          group                       group   \n",
       "1                Homo sapiens  MONDO_0005005  clear cell renal carcinoma   \n",
       "2                Homo sapiens  MONDO_0005005  clear cell renal carcinoma   \n",
       "3                Homo sapiens  MONDO_0005005  clear cell renal carcinoma   \n",
       "4                Homo sapiens  MONDO_0005005  clear cell renal carcinoma   \n",
       "...                       ...            ...                         ...   \n",
       "34322            Homo sapiens  MONDO_0005005  clear cell renal carcinoma   \n",
       "34323            Homo sapiens  MONDO_0005005  clear cell renal carcinoma   \n",
       "34324            Homo sapiens  MONDO_0005005  clear cell renal carcinoma   \n",
       "34325            Homo sapiens  MONDO_0005005  clear cell renal carcinoma   \n",
       "34326            Homo sapiens  MONDO_0005005  clear cell renal carcinoma   \n",
       "\n",
       "                organ organ__ontology_label library_preparation_protocol  \\\n",
       "0               group                 group                        group   \n",
       "1      UBERON_0000916               abdomen                  EFO:0009899   \n",
       "2      UBERON_0000916               abdomen                  EFO:0009899   \n",
       "3      UBERON_0000916               abdomen                  EFO:0009899   \n",
       "4      UBERON_0000916               abdomen                  EFO:0009899   \n",
       "...               ...                   ...                          ...   \n",
       "34322  UBERON_0000029            lymph node                  EFO:0009899   \n",
       "34323  UBERON_0000029            lymph node                  EFO:0009899   \n",
       "34324  UBERON_0000029            lymph node                  EFO:0009899   \n",
       "34325  UBERON_0000029            lymph node                  EFO:0009899   \n",
       "34326  UBERON_0000029            lymph node                  EFO:0009899   \n",
       "\n",
       "      library_preparation_protocol__ontology_label ICB_Exposed ICB_Response  \\\n",
       "0                                            group       group        group   \n",
       "1                             10X 3' v2 sequencing         ICB       ICB_PR   \n",
       "2                             10X 3' v2 sequencing         ICB       ICB_PR   \n",
       "3                             10X 3' v2 sequencing         ICB       ICB_PR   \n",
       "4                             10X 3' v2 sequencing         ICB       ICB_PR   \n",
       "...                                            ...         ...          ...   \n",
       "34322                         10X 3' v2 sequencing       NoICB        NoICB   \n",
       "34323                         10X 3' v2 sequencing       NoICB        NoICB   \n",
       "34324                         10X 3' v2 sequencing       NoICB        NoICB   \n",
       "34325                         10X 3' v2 sequencing       NoICB        NoICB   \n",
       "34326                         10X 3' v2 sequencing       NoICB        NoICB   \n",
       "\n",
       "      TKI_Exposed Initial_Louvain_Cluster   Lineage      InferCNV  \\\n",
       "0           group                   group     group         group   \n",
       "1             TKI                       9  Lymphoid  NonMalignant   \n",
       "2             TKI                       9  Lymphoid  NonMalignant   \n",
       "3             TKI                       9  Lymphoid  NonMalignant   \n",
       "4             TKI                       2  Lymphoid  NonMalignant   \n",
       "...           ...                     ...       ...           ...   \n",
       "34322       NoTKI                       7  Lymphoid  NonMalignant   \n",
       "34323       NoTKI                       0  Lymphoid  NonMalignant   \n",
       "34324       NoTKI                       9  Lymphoid  NonMalignant   \n",
       "34325       NoTKI                       7  Lymphoid  NonMalignant   \n",
       "34326       NoTKI                       9  Lymphoid  NonMalignant   \n",
       "\n",
       "             FinalCellType    sex  \n",
       "0                    group  group  \n",
       "1      41BB-Hi CD8+ T cell   male  \n",
       "2      41BB-Hi CD8+ T cell   male  \n",
       "3      41BB-Hi CD8+ T cell   male  \n",
       "4        MitoHigh T-Helper   male  \n",
       "...                    ...    ...  \n",
       "34322    Effector T-Helper   male  \n",
       "34323  41BB-Lo CD8+ T cell   male  \n",
       "34324  41BB-Hi CD8+ T cell   male  \n",
       "34325    Effector T-Helper   male  \n",
       "34326  41BB-Lo CD8+ T cell   male  \n",
       "\n",
       "[34327 rows x 19 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Load metadata\n",
    "entity = syn.get('syn59975076')\n",
    "metadata = pd.read_table(entity.path)\n",
    "\n",
    "metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "53914684-d3e4-4d6b-8155-2ce1f0b77f75",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['NAME', 'biosample_id', 'donor_id', 'species',\n",
       "       'species__ontology_label', 'disease', 'disease__ontology_label',\n",
       "       'organ', 'organ__ontology_label', 'library_preparation_protocol',\n",
       "       'library_preparation_protocol__ontology_label', 'ICB_Exposed',\n",
       "       'ICB_Response', 'TKI_Exposed', 'Initial_Louvain_Cluster', 'Lineage',\n",
       "       'InferCNV', 'FinalCellType', 'sex'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "be9daaab-de6b-4f4e-947f-cb505de78d51",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>biosample_id</th>\n",
       "      <th>donor_id</th>\n",
       "      <th>ICB_Exposed</th>\n",
       "      <th>ICB_Response</th>\n",
       "      <th>TKI_Exposed</th>\n",
       "      <th>sex</th>\n",
       "      <th>Lineage</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>group</td>\n",
       "      <td>group</td>\n",
       "      <td>group</td>\n",
       "      <td>group</td>\n",
       "      <td>group</td>\n",
       "      <td>group</td>\n",
       "      <td>group</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>P55_scRNA</td>\n",
       "      <td>P55</td>\n",
       "      <td>ICB</td>\n",
       "      <td>ICB_PR</td>\n",
       "      <td>TKI</td>\n",
       "      <td>male</td>\n",
       "      <td>Lymphoid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>P55_scRNA</td>\n",
       "      <td>P55</td>\n",
       "      <td>ICB</td>\n",
       "      <td>ICB_PR</td>\n",
       "      <td>TKI</td>\n",
       "      <td>male</td>\n",
       "      <td>Lymphoid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>P55_scRNA</td>\n",
       "      <td>P55</td>\n",
       "      <td>ICB</td>\n",
       "      <td>ICB_PR</td>\n",
       "      <td>TKI</td>\n",
       "      <td>male</td>\n",
       "      <td>Lymphoid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>P55_scRNA</td>\n",
       "      <td>P55</td>\n",
       "      <td>ICB</td>\n",
       "      <td>ICB_PR</td>\n",
       "      <td>TKI</td>\n",
       "      <td>male</td>\n",
       "      <td>Lymphoid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34322</th>\n",
       "      <td>P916_scRNA</td>\n",
       "      <td>P916</td>\n",
       "      <td>NoICB</td>\n",
       "      <td>NoICB</td>\n",
       "      <td>NoTKI</td>\n",
       "      <td>male</td>\n",
       "      <td>Lymphoid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34323</th>\n",
       "      <td>P916_scRNA</td>\n",
       "      <td>P916</td>\n",
       "      <td>NoICB</td>\n",
       "      <td>NoICB</td>\n",
       "      <td>NoTKI</td>\n",
       "      <td>male</td>\n",
       "      <td>Lymphoid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34324</th>\n",
       "      <td>P916_scRNA</td>\n",
       "      <td>P916</td>\n",
       "      <td>NoICB</td>\n",
       "      <td>NoICB</td>\n",
       "      <td>NoTKI</td>\n",
       "      <td>male</td>\n",
       "      <td>Lymphoid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34325</th>\n",
       "      <td>P916_scRNA</td>\n",
       "      <td>P916</td>\n",
       "      <td>NoICB</td>\n",
       "      <td>NoICB</td>\n",
       "      <td>NoTKI</td>\n",
       "      <td>male</td>\n",
       "      <td>Lymphoid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34326</th>\n",
       "      <td>P916_scRNA</td>\n",
       "      <td>P916</td>\n",
       "      <td>NoICB</td>\n",
       "      <td>NoICB</td>\n",
       "      <td>NoTKI</td>\n",
       "      <td>male</td>\n",
       "      <td>Lymphoid</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>34327 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      biosample_id donor_id ICB_Exposed ICB_Response TKI_Exposed    sex  \\\n",
       "0            group    group       group        group       group  group   \n",
       "1        P55_scRNA      P55         ICB       ICB_PR         TKI   male   \n",
       "2        P55_scRNA      P55         ICB       ICB_PR         TKI   male   \n",
       "3        P55_scRNA      P55         ICB       ICB_PR         TKI   male   \n",
       "4        P55_scRNA      P55         ICB       ICB_PR         TKI   male   \n",
       "...            ...      ...         ...          ...         ...    ...   \n",
       "34322   P916_scRNA     P916       NoICB        NoICB       NoTKI   male   \n",
       "34323   P916_scRNA     P916       NoICB        NoICB       NoTKI   male   \n",
       "34324   P916_scRNA     P916       NoICB        NoICB       NoTKI   male   \n",
       "34325   P916_scRNA     P916       NoICB        NoICB       NoTKI   male   \n",
       "34326   P916_scRNA     P916       NoICB        NoICB       NoTKI   male   \n",
       "\n",
       "        Lineage  \n",
       "0         group  \n",
       "1      Lymphoid  \n",
       "2      Lymphoid  \n",
       "3      Lymphoid  \n",
       "4      Lymphoid  \n",
       "...         ...  \n",
       "34322  Lymphoid  \n",
       "34323  Lymphoid  \n",
       "34324  Lymphoid  \n",
       "34325  Lymphoid  \n",
       "34326  Lymphoid  \n",
       "\n",
       "[34327 rows x 7 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Getting a closer look at the columns that seem to be of annotation we'll need\n",
    "clin_annot = pd.DataFrame({\n",
    "        'biosample_id': metadata[\"biosample_id\"],\n",
    "        'donor_id': metadata[\"donor_id\"],\n",
    "        'ICB_Exposed': metadata[\"ICB_Exposed\"],\n",
    "        'ICB_Response': metadata[\"ICB_Response\"],\n",
    "        'TKI_Exposed': metadata[\"TKI_Exposed\"],\n",
    "        'sex': metadata[\"sex\"],\n",
    "        'Lineage': metadata[\"Lineage\"]\n",
    "    }\n",
    ")\n",
    "clin_annot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "4cb4d333-30b2-4626-935f-1a90155539b7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "FinalCellType\n",
       "41BB-Lo CD8+ T cell      5420\n",
       "TP2                      4599\n",
       "TP1                      3324\n",
       "FOLR2-Hi TAM             1528\n",
       "MitoHigh CD8+ T cell     1482\n",
       "MitoHigh Myeloid         1407\n",
       "Effector T-Helper        1389\n",
       "GPNMB-Hi TAM             1382\n",
       "41BB-Hi CD8+ T cell      1321\n",
       "MitoHigh T-Helper        1316\n",
       "FGFBP2- NK               1306\n",
       "VSIR-Hi TAM              1070\n",
       "B cell                    962\n",
       "CD16- Monocyte            844\n",
       "NKT                       811\n",
       "T-Reg                     750\n",
       "Cycling CD8+ T cell       701\n",
       "LowLibSize Macrophage     672\n",
       "Memory T-Helper           579\n",
       "FGFBP2+ NK                493\n",
       "Plasma cell               463\n",
       "MitoHigh NK               446\n",
       "CD16+ Monocyte            313\n",
       "CD1C+ DC                  308\n",
       "Misc/Undetermined         278\n",
       "Endothelial               271\n",
       "CXCL10-Hi TAM             226\n",
       "Cycling TAM               175\n",
       "MX1-Hi CD8+ T cell        132\n",
       "Cycling Tumor             117\n",
       "CLEC9A+ DC                111\n",
       "Fibroblast                 91\n",
       "Mast cell                  39\n",
       "group                       1\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Inspect more closely unique values for cell type\n",
    "metadata[\"FinalCellType\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "8b71c759-3870-4288-a63e-49ec5bca6049",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>biosample_id</th>\n",
       "      <th>donor_id</th>\n",
       "      <th>ICB_Exposed</th>\n",
       "      <th>ICB_Response</th>\n",
       "      <th>TKI_Exposed</th>\n",
       "      <th>sex</th>\n",
       "      <th>Lineage</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>34327</td>\n",
       "      <td>34327</td>\n",
       "      <td>34327</td>\n",
       "      <td>34327</td>\n",
       "      <td>34327</td>\n",
       "      <td>34327</td>\n",
       "      <td>34327</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>unique</th>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>3</td>\n",
       "      <td>6</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>top</th>\n",
       "      <td>P90_scRNA</td>\n",
       "      <td>P90</td>\n",
       "      <td>ICB</td>\n",
       "      <td>NoICB</td>\n",
       "      <td>NoTKI</td>\n",
       "      <td>male</td>\n",
       "      <td>Lymphoid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>freq</th>\n",
       "      <td>8426</td>\n",
       "      <td>8426</td>\n",
       "      <td>17672</td>\n",
       "      <td>16654</td>\n",
       "      <td>23195</td>\n",
       "      <td>17988</td>\n",
       "      <td>17298</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       biosample_id donor_id ICB_Exposed ICB_Response TKI_Exposed    sex  \\\n",
       "count         34327    34327       34327        34327       34327  34327   \n",
       "unique            9        9           3            6           3      3   \n",
       "top       P90_scRNA      P90         ICB        NoICB       NoTKI   male   \n",
       "freq           8426     8426       17672        16654       23195  17988   \n",
       "\n",
       "         Lineage  \n",
       "count      34327  \n",
       "unique         5  \n",
       "top     Lymphoid  \n",
       "freq       17298  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clin_annot.describe(include='object')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "14e4a04f-a3bb-4759-b3ca-f2e1dc602f59",
   "metadata": {},
   "source": [
    "## Expression data\n",
    "\n",
    "The authors provide a table with normalized values after QC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "7ed7a051-737d-47a5-b8e0-6e79cc86c6be",
   "metadata": {},
   "outputs": [],
   "source": [
    "entity = syn.get('syn59976437')\n",
    "norm_df = pd.read_csv(entity.path, delimiter = \"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "a76669b4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GENE</th>\n",
       "      <th>AAACCTGAGAATAGGG.p55</th>\n",
       "      <th>AAACCTGAGGCTAGGT.p55</th>\n",
       "      <th>AAACCTGCACTGTGTA.p55</th>\n",
       "      <th>AAACCTGCAGTCCTTC.p55</th>\n",
       "      <th>AAACCTGGTAAATGTG.p55</th>\n",
       "      <th>AAACCTGGTACCGAGA.p55</th>\n",
       "      <th>AAACCTGGTGTGAAAT.p55</th>\n",
       "      <th>AAACCTGTCAGATAAG.p55</th>\n",
       "      <th>AAACCTGTCCTGCTTG.p55</th>\n",
       "      <th>...</th>\n",
       "      <th>TTCTACACAATCGGTT.p916</th>\n",
       "      <th>TTCTTAGGTCCATCCT.p916</th>\n",
       "      <th>TTGACTTAGAGCTATA.p916</th>\n",
       "      <th>TTGCCGTCACAAGCCC.p916</th>\n",
       "      <th>TTGCGTCCAAGTTAAG.p916</th>\n",
       "      <th>TTGGAACGTGAGGGAG.p916</th>\n",
       "      <th>TTGTAGGGTATGAAAC.p916</th>\n",
       "      <th>TTTACTGCACACATGT.p916</th>\n",
       "      <th>TTTGTCAAGAGCAATT.p916</th>\n",
       "      <th>TTTGTCAAGCGTTTAC.p916</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>WASH7P</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AL627309.1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AL627309.3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CICP27</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AL627309.1.1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32713</th>\n",
       "      <td>CU638689.5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32714</th>\n",
       "      <td>CU634019.2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32715</th>\n",
       "      <td>CU634019.6</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32716</th>\n",
       "      <td>CU638689.1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32717</th>\n",
       "      <td>FP671120.5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>32718 rows × 34327 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               GENE  AAACCTGAGAATAGGG.p55  AAACCTGAGGCTAGGT.p55  \\\n",
       "0            WASH7P                   0.0                   0.0   \n",
       "1        AL627309.1                   0.0                   0.0   \n",
       "2        AL627309.3                   0.0                   0.0   \n",
       "3            CICP27                   0.0                   0.0   \n",
       "4      AL627309.1.1                   0.0                   0.0   \n",
       "...             ...                   ...                   ...   \n",
       "32713    CU638689.5                   0.0                   0.0   \n",
       "32714    CU634019.2                   0.0                   0.0   \n",
       "32715    CU634019.6                   0.0                   0.0   \n",
       "32716    CU638689.1                   0.0                   0.0   \n",
       "32717    FP671120.5                   0.0                   0.0   \n",
       "\n",
       "       AAACCTGCACTGTGTA.p55  AAACCTGCAGTCCTTC.p55  AAACCTGGTAAATGTG.p55  \\\n",
       "0                       0.0                   0.0                   0.0   \n",
       "1                       0.0                   0.0                   0.0   \n",
       "2                       0.0                   0.0                   0.0   \n",
       "3                       0.0                   0.0                   0.0   \n",
       "4                       0.0                   0.0                   0.0   \n",
       "...                     ...                   ...                   ...   \n",
       "32713                   0.0                   0.0                   0.0   \n",
       "32714                   0.0                   0.0                   0.0   \n",
       "32715                   0.0                   0.0                   0.0   \n",
       "32716                   0.0                   0.0                   0.0   \n",
       "32717                   0.0                   0.0                   0.0   \n",
       "\n",
       "       AAACCTGGTACCGAGA.p55  AAACCTGGTGTGAAAT.p55  AAACCTGTCAGATAAG.p55  \\\n",
       "0                       0.0                   0.0                   0.0   \n",
       "1                       0.0                   0.0                   0.0   \n",
       "2                       0.0                   0.0                   0.0   \n",
       "3                       0.0                   0.0                   0.0   \n",
       "4                       0.0                   0.0                   0.0   \n",
       "...                     ...                   ...                   ...   \n",
       "32713                   0.0                   0.0                   0.0   \n",
       "32714                   0.0                   0.0                   0.0   \n",
       "32715                   0.0                   0.0                   0.0   \n",
       "32716                   0.0                   0.0                   0.0   \n",
       "32717                   0.0                   0.0                   0.0   \n",
       "\n",
       "       AAACCTGTCCTGCTTG.p55  ...  TTCTACACAATCGGTT.p916  \\\n",
       "0                       0.0  ...                    0.0   \n",
       "1                       0.0  ...                    0.0   \n",
       "2                       0.0  ...                    0.0   \n",
       "3                       0.0  ...                    0.0   \n",
       "4                       0.0  ...                    0.0   \n",
       "...                     ...  ...                    ...   \n",
       "32713                   0.0  ...                    0.0   \n",
       "32714                   0.0  ...                    0.0   \n",
       "32715                   0.0  ...                    0.0   \n",
       "32716                   0.0  ...                    0.0   \n",
       "32717                   0.0  ...                    0.0   \n",
       "\n",
       "       TTCTTAGGTCCATCCT.p916  TTGACTTAGAGCTATA.p916  TTGCCGTCACAAGCCC.p916  \\\n",
       "0                        0.0                    0.0                    0.0   \n",
       "1                        0.0                    0.0                    0.0   \n",
       "2                        0.0                    0.0                    0.0   \n",
       "3                        0.0                    0.0                    0.0   \n",
       "4                        0.0                    0.0                    0.0   \n",
       "...                      ...                    ...                    ...   \n",
       "32713                    0.0                    0.0                    0.0   \n",
       "32714                    0.0                    0.0                    0.0   \n",
       "32715                    0.0                    0.0                    0.0   \n",
       "32716                    0.0                    0.0                    0.0   \n",
       "32717                    0.0                    0.0                    0.0   \n",
       "\n",
       "       TTGCGTCCAAGTTAAG.p916  TTGGAACGTGAGGGAG.p916  TTGTAGGGTATGAAAC.p916  \\\n",
       "0                        0.0                    0.0                    0.0   \n",
       "1                        0.0                    0.0                    0.0   \n",
       "2                        0.0                    0.0                    0.0   \n",
       "3                        0.0                    0.0                    0.0   \n",
       "4                        0.0                    0.0                    0.0   \n",
       "...                      ...                    ...                    ...   \n",
       "32713                    0.0                    0.0                    0.0   \n",
       "32714                    0.0                    0.0                    0.0   \n",
       "32715                    0.0                    0.0                    0.0   \n",
       "32716                    0.0                    0.0                    0.0   \n",
       "32717                    0.0                    0.0                    0.0   \n",
       "\n",
       "       TTTACTGCACACATGT.p916  TTTGTCAAGAGCAATT.p916  TTTGTCAAGCGTTTAC.p916  \n",
       "0                        0.0                    0.0                    0.0  \n",
       "1                        0.0                    0.0                    0.0  \n",
       "2                        0.0                    0.0                    0.0  \n",
       "3                        0.0                    0.0                    0.0  \n",
       "4                        0.0                    0.0                    0.0  \n",
       "...                      ...                    ...                    ...  \n",
       "32713                    0.0                    0.0                    0.0  \n",
       "32714                    0.0                    0.0                    0.0  \n",
       "32715                    0.0                    0.0                    0.0  \n",
       "32716                    0.0                    0.0                    0.0  \n",
       "32717                    0.0                    0.0                    0.0  \n",
       "\n",
       "[32718 rows x 34327 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "norm_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b7630a12",
   "metadata": {},
   "source": [
    "This normalization, according to the original paper was made using Seurat, with log2. Other single cell datasets in iAtlas were normalized using the scanpy.pp.log1p function, so ideally we will process the un-normalized data.\n",
    "\n",
    "The authors provide a matrix with raw counts."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "987beea4-15ff-44c2-bc63-440fd06502df",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<60627x39391 sparse matrix of type '<class 'numpy.int64'>'\n",
       "\twith 46152142 stored elements in COOrdinate format>"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "entity = syn.get('syn60521660')\n",
    "a = mmread(entity.path)\n",
    "a"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "f16bc57c-e591-4fa0-8f9f-afd563660e9e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "matrix([[0, 0, 0, ..., 0, 0, 0],\n",
       "        [0, 0, 0, ..., 0, 0, 0],\n",
       "        [0, 0, 0, ..., 0, 0, 0],\n",
       "        ...,\n",
       "        [0, 0, 0, ..., 0, 0, 0],\n",
       "        [0, 0, 0, ..., 0, 0, 0],\n",
       "        [0, 0, 0, ..., 0, 0, 0]])"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "a.todense()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "eb2e9ff7-b18e-4972-a337-28a765e05143",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[WARNING] /home/kawakami/miniconda3/envs/pvision/lib/python3.7/site-packages/ipykernel_launcher.py:2: FutureWarning: X.dtype being converted to np.float32 from int64. In the next version of anndata (0.9) conversion will not be automatic. Pass dtype explicitly to avoid this warning. Pass `AnnData(X, dtype=X.dtype, ...)` to get the future behavour.\n",
      "  \n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 39391 × 60627"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# creating Anndata file\n",
    "adata = ad.AnnData(sps.coo_matrix.transpose(a).todense())\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "503173c7-f406-4553-a16d-50e7cbce76c4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(39391, 1)"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "entity = syn.get('syn60521658')\n",
    "obs_names = pd.read_table(entity.path, header = None, names = [\"NAME\"])\n",
    "obs_names.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "40c60c12-686f-41f1-acba-bb65d33aa5e0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>symbol</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>DDX11L1</th>\n",
       "      <td>DDX11L1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WASH7P</th>\n",
       "      <td>WASH7P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIR1302-2HG</th>\n",
       "      <td>MIR1302-2HG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FAM138A</th>\n",
       "      <td>FAM138A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OR4G4P</th>\n",
       "      <td>OR4G4P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CU638689.1</th>\n",
       "      <td>CU638689.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FP236383.3</th>\n",
       "      <td>FP236383.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CU633904.2</th>\n",
       "      <td>CU633904.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FP671120.5</th>\n",
       "      <td>FP671120.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC233701.1</th>\n",
       "      <td>AC233701.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>60627 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  symbol\n",
       "gene                    \n",
       "DDX11L1          DDX11L1\n",
       "WASH7P            WASH7P\n",
       "MIR1302-2HG  MIR1302-2HG\n",
       "FAM138A          FAM138A\n",
       "OR4G4P            OR4G4P\n",
       "...                  ...\n",
       "CU638689.1    CU638689.1\n",
       "FP236383.3    FP236383.3\n",
       "CU633904.2    CU633904.2\n",
       "FP671120.5    FP671120.5\n",
       "AC233701.1    AC233701.1\n",
       "\n",
       "[60627 rows x 1 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "entity = syn.get('syn60521659')\n",
    "var_names = pd.read_table(entity.path, header = None, index_col = 0, names = [\"gene\", \"symbol\"])\n",
    "var_names.shape\n",
    "var_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "51122a14-39ad-440b-bf54-01dc44408df9",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs = obs_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "b72ddf0e-17ee-4782-9e11-9925d5ca0173",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['AAACCTGAGAATAGGG.p55', 'AAACCTGAGGCTAGGT.p55', 'AAACCTGCACTGTGTA.p55',\n",
       "       'AAACCTGCAGTCCTTC.p55', 'AAACCTGGTAAATGTG.p55', 'AAACCTGGTACCGAGA.p55',\n",
       "       'AAACCTGGTGTGAAAT.p55', 'AAACCTGTCAGATAAG.p55', 'AAACCTGTCCTGCTTG.p55',\n",
       "       'AAACCTGTCGCAAGCC.p55',\n",
       "       ...\n",
       "       'TTCTTAGGTCCATCCT.p916', 'TTGACTTAGAGCTATA.p916',\n",
       "       'TTGCCGTCACAAGCCC.p916', 'TTGCGTCCAAGTTAAG.p916',\n",
       "       'TTGGAACGTGAGGGAG.p916', 'TTGTAGGGTATGAAAC.p916',\n",
       "       'TTTACTGCACACATGT.p916', 'TTTATGCGTGAAAGAG.p916',\n",
       "       'TTTGTCAAGAGCAATT.p916', 'TTTGTCAAGCGTTTAC.p916'],\n",
       "      dtype='object', name='NAME', length=39391)"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs_names = obs_names.iloc[:,0]\n",
    "adata.obs_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "e9416675-a3e1-4a54-af1a-74c87fb42f18",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['DDX11L1', 'WASH7P', 'MIR1302-2HG', 'FAM138A', 'OR4G4P', 'OR4G11P',\n",
       "       'OR4F5', 'AL627309.1', 'AL627309.3', 'CICP27',\n",
       "       ...\n",
       "       'CU633904.3', 'CU634019.2', 'CU638689.3', 'CU634019.6', 'CU634019.4',\n",
       "       'CU638689.1', 'FP236383.3', 'CU633904.2', 'FP671120.5', 'AC233701.1'],\n",
       "      dtype='object', name='symbol', length=60627)"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.var = var_names\n",
    "adata.var_names = var_names.iloc[:,0]\n",
    "adata.var_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "463ae43c-9185-4550-960e-5c819082461a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 39391 × 60627\n",
       "    obs: 'NAME'\n",
       "    var: 'symbol'"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3767d5e4-aa47-4488-b344-7003d4ede61f",
   "metadata": {},
   "source": [
    "The number of cells in the counts matrix is higher than the one in the metadata and normalized files. The assumption is that the matadata file has the cells after QC. In the next section we will explore the available clinical data and then add this to the anndata object."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4f18bde0",
   "metadata": {},
   "source": [
    "# QC"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3abd2dd1",
   "metadata": {},
   "source": [
    "The authors didn't share QC metrics in the data. In the manuscript, they share the following filtering steps:\n",
    "\n",
    "*To exclude data from droplets containing more than one cell, doublet detection and removal were performed on gene-barcode matrices using Scrublet (Wolock et al., 2019). An expected doublet rate parameter of 0.06 was used, and doublet score thresholds were chosen manually to divide putative singlet and neotypic doublet modes in the score distribution. Predicted doublets were then removed from gene-barcode matrices. Prior to across-sample integration, cells with fewer than 200 genes detected or more than 25% of counts attributed to mitochondrially-encoded transcripts were removed. Genes detected in fewer than three cells across all samples were also excluded.\n",
    "The resulting merged dataset included 34,326 cells and 32,718 detected genes across the eight samples of the cohort.*\n",
    "\n",
    "We can easily compute the mitochondrial fraction and number of counts per cell. However, the manual steps are not esily reproducible.\n",
    "The number of cells and genes in norm_df are the same as mentioned in the manuscript as being the ones selected after QC. We will use them to filter cells and genes in the anndata object and assume we have the correct output from QC."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "48a384c3-0906-4044-828d-652b3b9da4bb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "View of AnnData object with n_obs × n_vars = 34326 × 32636\n",
       "    obs: 'NAME'\n",
       "    var: 'symbol'"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "qcdata = adata[:, np.isin(adata.var[\"symbol\"], norm_df[\"GENE\"])]\n",
    "qcdata = qcdata[np.isin(adata.obs[\"NAME\"], metadata[\"NAME\"])]\n",
    "qcdata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b412600e-e708-4ba5-bcad-8da3e94dba51",
   "metadata": {},
   "source": [
    "### Counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "4c7502b6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>symbol</th>\n",
       "      <th>WASH7P</th>\n",
       "      <th>AL627309.1</th>\n",
       "      <th>AL627309.3</th>\n",
       "      <th>CICP27</th>\n",
       "      <th>AL627309.1.1</th>\n",
       "      <th>AL627309.5</th>\n",
       "      <th>AP006222.2</th>\n",
       "      <th>RP4-669L17.8</th>\n",
       "      <th>AL669831.3</th>\n",
       "      <th>MTND1P23</th>\n",
       "      <th>...</th>\n",
       "      <th>LINC01670</th>\n",
       "      <th>FP236241.1</th>\n",
       "      <th>CU633906.1</th>\n",
       "      <th>CU633967.1</th>\n",
       "      <th>CU634019.1</th>\n",
       "      <th>CU638689.5</th>\n",
       "      <th>CU634019.2</th>\n",
       "      <th>CU634019.6</th>\n",
       "      <th>CU638689.1</th>\n",
       "      <th>FP671120.5</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NAME</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGAATAGGG.p55</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGGCTAGGT.p55</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGCACTGTGTA.p55</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGCAGTCCTTC.p55</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGGTAAATGTG.p55</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTGGAACGTGAGGGAG.p916</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTGTAGGGTATGAAAC.p916</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTACTGCACACATGT.p916</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCAAGAGCAATT.p916</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCAAGCGTTTAC.p916</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>34326 rows × 32636 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "symbol                 WASH7P  AL627309.1  AL627309.3  CICP27  AL627309.1.1  \\\n",
       "NAME                                                                          \n",
       "AAACCTGAGAATAGGG.p55      0.0         0.0         0.0     0.0           0.0   \n",
       "AAACCTGAGGCTAGGT.p55      0.0         0.0         0.0     0.0           0.0   \n",
       "AAACCTGCACTGTGTA.p55      0.0         0.0         0.0     0.0           0.0   \n",
       "AAACCTGCAGTCCTTC.p55      0.0         0.0         0.0     0.0           0.0   \n",
       "AAACCTGGTAAATGTG.p55      0.0         0.0         0.0     0.0           0.0   \n",
       "...                       ...         ...         ...     ...           ...   \n",
       "TTGGAACGTGAGGGAG.p916     0.0         0.0         0.0     0.0           0.0   \n",
       "TTGTAGGGTATGAAAC.p916     0.0         0.0         0.0     0.0           0.0   \n",
       "TTTACTGCACACATGT.p916     0.0         0.0         0.0     0.0           0.0   \n",
       "TTTGTCAAGAGCAATT.p916     0.0         0.0         0.0     0.0           0.0   \n",
       "TTTGTCAAGCGTTTAC.p916     0.0         0.0         0.0     0.0           0.0   \n",
       "\n",
       "symbol                 AL627309.5  AP006222.2  RP4-669L17.8  AL669831.3  \\\n",
       "NAME                                                                      \n",
       "AAACCTGAGAATAGGG.p55          0.0         0.0           0.0         0.0   \n",
       "AAACCTGAGGCTAGGT.p55          0.0         0.0           0.0         0.0   \n",
       "AAACCTGCACTGTGTA.p55          0.0         0.0           0.0         0.0   \n",
       "AAACCTGCAGTCCTTC.p55          0.0         0.0           0.0         0.0   \n",
       "AAACCTGGTAAATGTG.p55          0.0         0.0           0.0         0.0   \n",
       "...                           ...         ...           ...         ...   \n",
       "TTGGAACGTGAGGGAG.p916         0.0         0.0           0.0         0.0   \n",
       "TTGTAGGGTATGAAAC.p916         0.0         0.0           0.0         0.0   \n",
       "TTTACTGCACACATGT.p916         0.0         0.0           0.0         0.0   \n",
       "TTTGTCAAGAGCAATT.p916         0.0         0.0           0.0         0.0   \n",
       "TTTGTCAAGCGTTTAC.p916         0.0         0.0           0.0         0.0   \n",
       "\n",
       "symbol                 MTND1P23  ...  LINC01670  FP236241.1  CU633906.1  \\\n",
       "NAME                             ...                                      \n",
       "AAACCTGAGAATAGGG.p55        0.0  ...        0.0         0.0         0.0   \n",
       "AAACCTGAGGCTAGGT.p55        0.0  ...        0.0         0.0         0.0   \n",
       "AAACCTGCACTGTGTA.p55        0.0  ...        0.0         0.0         0.0   \n",
       "AAACCTGCAGTCCTTC.p55        0.0  ...        0.0         0.0         0.0   \n",
       "AAACCTGGTAAATGTG.p55        0.0  ...        0.0         0.0         0.0   \n",
       "...                         ...  ...        ...         ...         ...   \n",
       "TTGGAACGTGAGGGAG.p916       0.0  ...        0.0         0.0         0.0   \n",
       "TTGTAGGGTATGAAAC.p916       0.0  ...        0.0         0.0         0.0   \n",
       "TTTACTGCACACATGT.p916       0.0  ...        0.0         0.0         0.0   \n",
       "TTTGTCAAGAGCAATT.p916       0.0  ...        0.0         0.0         0.0   \n",
       "TTTGTCAAGCGTTTAC.p916       0.0  ...        0.0         0.0         0.0   \n",
       "\n",
       "symbol                 CU633967.1  CU634019.1  CU638689.5  CU634019.2  \\\n",
       "NAME                                                                    \n",
       "AAACCTGAGAATAGGG.p55          0.0         0.0         0.0         0.0   \n",
       "AAACCTGAGGCTAGGT.p55          0.0         0.0         0.0         0.0   \n",
       "AAACCTGCACTGTGTA.p55          0.0         0.0         0.0         0.0   \n",
       "AAACCTGCAGTCCTTC.p55          0.0         0.0         0.0         0.0   \n",
       "AAACCTGGTAAATGTG.p55          0.0         0.0         0.0         0.0   \n",
       "...                           ...         ...         ...         ...   \n",
       "TTGGAACGTGAGGGAG.p916         0.0         0.0         0.0         0.0   \n",
       "TTGTAGGGTATGAAAC.p916         0.0         0.0         0.0         0.0   \n",
       "TTTACTGCACACATGT.p916         0.0         0.0         0.0         0.0   \n",
       "TTTGTCAAGAGCAATT.p916         0.0         0.0         0.0         0.0   \n",
       "TTTGTCAAGCGTTTAC.p916         0.0         0.0         0.0         0.0   \n",
       "\n",
       "symbol                 CU634019.6  CU638689.1  FP671120.5  \n",
       "NAME                                                       \n",
       "AAACCTGAGAATAGGG.p55          0.0         0.0         0.0  \n",
       "AAACCTGAGGCTAGGT.p55          0.0         0.0         0.0  \n",
       "AAACCTGCACTGTGTA.p55          0.0         0.0         0.0  \n",
       "AAACCTGCAGTCCTTC.p55          0.0         0.0         0.0  \n",
       "AAACCTGGTAAATGTG.p55          0.0         0.0         0.0  \n",
       "...                           ...         ...         ...  \n",
       "TTGGAACGTGAGGGAG.p916         0.0         0.0         0.0  \n",
       "TTGTAGGGTATGAAAC.p916         0.0         0.0         0.0  \n",
       "TTTACTGCACACATGT.p916         0.0         0.0         0.0  \n",
       "TTTGTCAAGAGCAATT.p916         0.0         0.0         0.0  \n",
       "TTTGTCAAGCGTTTAC.p916         0.0         0.0         0.0  \n",
       "\n",
       "[34326 rows x 32636 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#saving counts as a layer\n",
    "qcdata.layers['counts'] = qcdata.X\n",
    "qcdata.to_df(layer = \"counts\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d02d6839",
   "metadata": {},
   "source": [
    "# Formatting to run pseudobulk"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "14ae77d0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Normalize and log-transform\n",
    "sc.pp.normalize_total(qcdata, target_sum=1e4)\n",
    "sc.pp.log1p(qcdata)\n",
    "qcdata.layers['normalized'] = qcdata.X\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "15f5d2ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "qcdata.obs.reset_index(drop = True, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "38233c7b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[WARNING] /home/kawakami/miniconda3/envs/pvision/lib/python3.7/site-packages/anndata/_core/anndata.py:798: UserWarning: \n",
      "AnnData expects .obs.index to contain strings, but got values like:\n",
      "    [0, 1, 2, 3, 4]\n",
      "\n",
      "    Inferred to be: integer\n",
      "\n",
      "  value_idx = self._prep_dim_index(value.index, attr)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# merging the metadata\n",
    "qcdata.obs = pd.merge(qcdata.obs, metadata.iloc[1:].drop('Initial_Louvain_Cluster', axis=1), on = \"NAME\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "b6e3b13d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 34326 × 32636\n",
       "    obs: 'NAME', 'biosample_id', 'donor_id', 'species', 'species__ontology_label', 'disease', 'disease__ontology_label', 'organ', 'organ__ontology_label', 'library_preparation_protocol', 'library_preparation_protocol__ontology_label', 'ICB_Exposed', 'ICB_Response', 'TKI_Exposed', 'Lineage', 'InferCNV', 'FinalCellType', 'sex'\n",
       "    var: 'symbol'\n",
       "    uns: 'log1p'\n",
       "    layers: 'counts', 'normalized'"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "qcdata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "410beb34-e346-4772-9594-c417c1cac146",
   "metadata": {},
   "outputs": [],
   "source": [
    "#This dataset has different names for cells than what we will use in iAtlas, so let's update them\n",
    "def func(a):\n",
    "    if a in [\"41BB-Lo CD8+ T cell\", \"MX1-Hi CD8+ T cell\", \"Memory T-Helper\", \"Cycling CD8+ T cell\", \"T-Reg\", \"MitoHigh T-Helper\", \"41BB-Hi CD8+ T cell\", \"Effector T-Helper\", \"MitoHigh CD8+ T cell\"]:\n",
    "        return \"T cell\"\n",
    "    elif a in [\"LowLibSize Macrophage\", \"VSIR-Hi TAM\", \"Cycling TAM\", \"CXCL10-Hi TAM\", \"GPNMB-Hi TAM\", \"FOLR2-Hi TAM\"]:\n",
    "        return \"macrophage\"\n",
    "    elif a in \"MitoHigh Myeloid\":\n",
    "        return \"myeloid cell\"\n",
    "    elif a in \"B cell\":\n",
    "        return \"B cell\"\n",
    "    elif a in \"Mast cell\":\n",
    "        return \"mast cell\"\n",
    "    elif a in \"Endothelial\":\n",
    "        return \"endothelium\"\n",
    "    elif a in \"Fibroblast\":\n",
    "        return \"fibroblast\"\n",
    "    elif a in [\"TP1\", \"TP2\", \"Cycling Tumor\"]:\n",
    "        return \"tumor\"\n",
    "    elif a in \"Plasma cell\":\n",
    "        return \"plasma cell\"\n",
    "    elif a in [\"CD16- Monocyte\", \"CD16+ Monocyte\"]:\n",
    "        return \"monocyte\"\n",
    "    elif a in [\"MitoHigh NK\", \"FGFBP2+ NK\", \"NKT\", \"FGFBP2- NK\"]:\n",
    "        return \"NK\"\n",
    "    elif a in [\"CLEC9A+ DC\", \"CD1C+ DC\"]:\n",
    "        return \"Dendritic cell\"\n",
    "    else:\n",
    "        return a\n",
    "\n",
    "qcdata.obs[\"cell_type_iatlas\"] = qcdata.obs[\"FinalCellType\"].apply(lambda x: func(x))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "34e95f61-ccd1-4db7-9891-0cc6a1af464c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "T cell               13090\n",
       "tumor                 8040\n",
       "macrophage            5053\n",
       "NK                    3056\n",
       "myeloid cell          1407\n",
       "monocyte              1157\n",
       "B cell                 962\n",
       "plasma cell            463\n",
       "Dendritic cell         419\n",
       "Misc/Undetermined      278\n",
       "endothelium            271\n",
       "fibroblast              91\n",
       "mast cell               39\n",
       "Name: cell_type_iatlas, dtype: int64"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "qcdata.obs[\"cell_type_iatlas\"].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b5a38eaa-ba22-4a9f-bdb0-801b34d463b7",
   "metadata": {},
   "source": [
    "# File uploads\n",
    "\n",
    "### h5ad file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "d2044141",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "##################################################\n",
      " Uploading file to Synapse storage \n",
      "##################################################\n",
      "\n"
     ]
    }
   ],
   "source": [
    "qcdata.write('bi_iatlas.h5ad')\n",
    "file_entity = synapseclient.File('bi_iatlas.h5ad', 'syn59966587')\n",
    "file_entity = syn.store(file_entity)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "118d79d9-7037-44a6-94a2-0964b2451020",
   "metadata": {},
   "source": [
    "### UMAP coordinates\n",
    "\n",
    "They are stored in a separate file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "95c0e6b0-6b0a-4f16-be75-012125732692",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>NAME</th>\n",
       "      <th>X</th>\n",
       "      <th>Y</th>\n",
       "      <th>FinalCellType</th>\n",
       "      <th>Initial_Louvain_Cluster</th>\n",
       "      <th>Lineage</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAACCTGAGAATAGGG.p55</td>\n",
       "      <td>-6.02505302</td>\n",
       "      <td>-7.725045658</td>\n",
       "      <td>41BB-Hi CD8+ T cell</td>\n",
       "      <td>9</td>\n",
       "      <td>Lymphoid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAACCTGAGGCTAGGT.p55</td>\n",
       "      <td>-6.357606407</td>\n",
       "      <td>-7.995055652</td>\n",
       "      <td>41BB-Hi CD8+ T cell</td>\n",
       "      <td>9</td>\n",
       "      <td>Lymphoid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAACCTGCACTGTGTA.p55</td>\n",
       "      <td>-6.564451214</td>\n",
       "      <td>-8.440553165</td>\n",
       "      <td>41BB-Hi CD8+ T cell</td>\n",
       "      <td>9</td>\n",
       "      <td>Lymphoid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAACCTGCAGTCCTTC.p55</td>\n",
       "      <td>-0.524987217</td>\n",
       "      <td>-3.175035215</td>\n",
       "      <td>MitoHigh T-Helper</td>\n",
       "      <td>2</td>\n",
       "      <td>Lymphoid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>AAACCTGGTAAATGTG.p55</td>\n",
       "      <td>-5.791819092</td>\n",
       "      <td>-5.585831619</td>\n",
       "      <td>41BB-Lo CD8+ T cell</td>\n",
       "      <td>0</td>\n",
       "      <td>Lymphoid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34322</th>\n",
       "      <td>TTGGAACGTGAGGGAG.p916</td>\n",
       "      <td>-5.671237465</td>\n",
       "      <td>-2.7540519</td>\n",
       "      <td>Effector T-Helper</td>\n",
       "      <td>7</td>\n",
       "      <td>Lymphoid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34323</th>\n",
       "      <td>TTGTAGGGTATGAAAC.p916</td>\n",
       "      <td>-6.444504257</td>\n",
       "      <td>-7.248266674</td>\n",
       "      <td>41BB-Lo CD8+ T cell</td>\n",
       "      <td>0</td>\n",
       "      <td>Lymphoid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34324</th>\n",
       "      <td>TTTACTGCACACATGT.p916</td>\n",
       "      <td>-6.708637711</td>\n",
       "      <td>-8.686034656</td>\n",
       "      <td>41BB-Hi CD8+ T cell</td>\n",
       "      <td>9</td>\n",
       "      <td>Lymphoid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34325</th>\n",
       "      <td>TTTGTCAAGAGCAATT.p916</td>\n",
       "      <td>-5.968340393</td>\n",
       "      <td>-2.565747714</td>\n",
       "      <td>Effector T-Helper</td>\n",
       "      <td>7</td>\n",
       "      <td>Lymphoid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34326</th>\n",
       "      <td>TTTGTCAAGCGTTTAC.p916</td>\n",
       "      <td>-6.146290775</td>\n",
       "      <td>-8.060590244</td>\n",
       "      <td>41BB-Lo CD8+ T cell</td>\n",
       "      <td>9</td>\n",
       "      <td>Lymphoid</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>34326 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        NAME             X             Y        FinalCellType  \\\n",
       "1       AAACCTGAGAATAGGG.p55   -6.02505302  -7.725045658  41BB-Hi CD8+ T cell   \n",
       "2       AAACCTGAGGCTAGGT.p55  -6.357606407  -7.995055652  41BB-Hi CD8+ T cell   \n",
       "3       AAACCTGCACTGTGTA.p55  -6.564451214  -8.440553165  41BB-Hi CD8+ T cell   \n",
       "4       AAACCTGCAGTCCTTC.p55  -0.524987217  -3.175035215    MitoHigh T-Helper   \n",
       "5       AAACCTGGTAAATGTG.p55  -5.791819092  -5.585831619  41BB-Lo CD8+ T cell   \n",
       "...                      ...           ...           ...                  ...   \n",
       "34322  TTGGAACGTGAGGGAG.p916  -5.671237465    -2.7540519    Effector T-Helper   \n",
       "34323  TTGTAGGGTATGAAAC.p916  -6.444504257  -7.248266674  41BB-Lo CD8+ T cell   \n",
       "34324  TTTACTGCACACATGT.p916  -6.708637711  -8.686034656  41BB-Hi CD8+ T cell   \n",
       "34325  TTTGTCAAGAGCAATT.p916  -5.968340393  -2.565747714    Effector T-Helper   \n",
       "34326  TTTGTCAAGCGTTTAC.p916  -6.146290775  -8.060590244  41BB-Lo CD8+ T cell   \n",
       "\n",
       "      Initial_Louvain_Cluster   Lineage  \n",
       "1                           9  Lymphoid  \n",
       "2                           9  Lymphoid  \n",
       "3                           9  Lymphoid  \n",
       "4                           2  Lymphoid  \n",
       "5                           0  Lymphoid  \n",
       "...                       ...       ...  \n",
       "34322                       7  Lymphoid  \n",
       "34323                       0  Lymphoid  \n",
       "34324                       9  Lymphoid  \n",
       "34325                       7  Lymphoid  \n",
       "34326                       9  Lymphoid  \n",
       "\n",
       "[34326 rows x 6 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "umap = pd.read_table(\"/Users/heimann/Documents/cri_iatlas/htan-dev/Bi_2021/SCP1288/cluster/Final_SCP_ClusterFile.txt\")\n",
    "umap_df = umap.drop(index=0) #first row has metadata about column, exclude it\n",
    "umap_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "3986eb5b-7e2c-4d3a-9104-8393c16c49e3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>umap_1</th>\n",
       "      <th>umap_2</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NAME</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGAATAGGG.p55</th>\n",
       "      <td>-6.02505302</td>\n",
       "      <td>-7.725045658</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGGCTAGGT.p55</th>\n",
       "      <td>-6.357606407</td>\n",
       "      <td>-7.995055652</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGCACTGTGTA.p55</th>\n",
       "      <td>-6.564451214</td>\n",
       "      <td>-8.440553165</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGCAGTCCTTC.p55</th>\n",
       "      <td>-0.524987217</td>\n",
       "      <td>-3.175035215</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGGTAAATGTG.p55</th>\n",
       "      <td>-5.791819092</td>\n",
       "      <td>-5.585831619</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTGGAACGTGAGGGAG.p916</th>\n",
       "      <td>-5.671237465</td>\n",
       "      <td>-2.7540519</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTGTAGGGTATGAAAC.p916</th>\n",
       "      <td>-6.444504257</td>\n",
       "      <td>-7.248266674</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTACTGCACACATGT.p916</th>\n",
       "      <td>-6.708637711</td>\n",
       "      <td>-8.686034656</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCAAGAGCAATT.p916</th>\n",
       "      <td>-5.968340393</td>\n",
       "      <td>-2.565747714</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCAAGCGTTTAC.p916</th>\n",
       "      <td>-6.146290775</td>\n",
       "      <td>-8.060590244</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>34326 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                             umap_1        umap_2\n",
       "NAME                                             \n",
       "AAACCTGAGAATAGGG.p55    -6.02505302  -7.725045658\n",
       "AAACCTGAGGCTAGGT.p55   -6.357606407  -7.995055652\n",
       "AAACCTGCACTGTGTA.p55   -6.564451214  -8.440553165\n",
       "AAACCTGCAGTCCTTC.p55   -0.524987217  -3.175035215\n",
       "AAACCTGGTAAATGTG.p55   -5.791819092  -5.585831619\n",
       "...                             ...           ...\n",
       "TTGGAACGTGAGGGAG.p916  -5.671237465    -2.7540519\n",
       "TTGTAGGGTATGAAAC.p916  -6.444504257  -7.248266674\n",
       "TTTACTGCACACATGT.p916  -6.708637711  -8.686034656\n",
       "TTTGTCAAGAGCAATT.p916  -5.968340393  -2.565747714\n",
       "TTTGTCAAGCGTTTAC.p916  -6.146290775  -8.060590244\n",
       "\n",
       "[34326 rows x 2 columns]"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#preparing UMAP file to be used in iAtlas\n",
    "umap_df.index = umap_df[\"NAME\"]\n",
    "umap_df = umap_df.rename(columns = {\"X\" : \"umap_1\", \"Y\":\"umap_2\"})\n",
    "umap_df[[\"umap_1\", \"umap_2\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "95842c86-ae49-47fb-acd6-07dfaccde8fa",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Uploading to Synapse storage: 100%|█| 1.59M/1.59M [00:01<00:00, 1.35MB/s, Bi_202\n"
     ]
    }
   ],
   "source": [
    "umap_df[[\"umap_1\", \"umap_2\"]].to_csv('Bi_2021_umap.tsv', sep='\\t', index=True)\n",
    "file_entity = synapseclient.File('Bi_2021_umap.tsv', 'syn59966566')\n",
    "file_entity = syn.store(file_entity)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
